### PR TITLE
Clean up invokable SettingsManagers and hierarchy

### PR DESCRIPTION
**Pull request by spbrogan**
Make APIs more consistent and don't override functions in Settings Managers.
Raise Exceptions for required functions
Set safe defaults for optional functions

### CODE REVIEW

**Comment by spbrogan** [comment]
Need to fix tests and understand why failing.  My goal was to refactor without breaking anything.  :)

File: edk2toolext/edk2_invocable.py [hunk 1]
```diff
@@ -22,6 +23,44 @@
...
+        txt  == plain text file logging
+        md   == markdown file logging
+        '''
+        return None
+
+    def AddCommandLineOptions(self, parserObj) -> None:
+        ''' Implement in subclass to add command line options to the argparser '''
+        pass
+
+    def RetrieveCommandLineOptions(self, args):
```
Comment (matthewfcarlson) [inline edk2toolext/edk2_invocable.py:59]: args would be an object (https://docs.python.org/3/library/typing.html)

File: edk2toolext/edk2_invocable.py [hunk 2]
```diff
@@ -22,6 +23,44 @@
...
+        This is an interface definition only
+        to show which functions are required to be implemented
+        and can be implemented in a settings manager.
+    '''
+
+    def GetWorkspaceRoot(self):
+        ''' get absolute path to WorkspaceRoot '''
+        raise NotImplementedError()
+
+    def GetPackagesPath(self) -> Iterable:
```
Comment (matthewfcarlson) [inline edk2toolext/edk2_invocable.py:38]: Iterable[os.PathLike] would be more specific but not required

File: edk2toolext/edk2_invocable.py [hunk 3]
```diff
@@ -22,6 +23,44 @@
...
+        ''' Optional API to return an Iterable of paths that should be mapped as edk2 PackagesPath '''
+        return []
+
+    def GetActiveScopes(self) -> Tuple(str):
```
Comment (matthewfcarlson) [inline edk2toolext/edk2_invocable.py:42]: Tuple[str]

File: edk2toolext/edk2_invocable.py (hunk 1 above)
Comment (matthewfcarlson) [inline edk2toolext/edk2_invocable.py:59]: -> None

File: edk2toolext/edk2_invocable.py [hunk 4]
```diff
@@ -46,6 +85,12 @@ def GetWorkspaceRoot(self):
         except AttributeError:
             raise RuntimeError("Can't call this before PlatformSettings has been set up!")
 
+    def GetPackagesPath(self):
+        try:
+            return self.PlatformSettings.GetPackagesPath()
+        except AttributeError:
+            raise RuntimeError("Can't call this before PlatformSettings has been set up!")
+
     def GetActiveScopes(self):
```
Comment (matthewfcarlson) [inline edk2toolext/edk2_invocable.py:94]: -> Iterable[str]

Comment (matthewfcarlson) [reply edk2toolext/edk2_invocable.py:94]: Do you need a tuple or iterable?

File: edk2toolext/invocables/edk2_ci_build.py [hunk 5]
```diff
@@ -25,68 +26,16 @@
...
-        ''' Get the logging level for a given type
-        base == lowest logging level supported
-        con  == Screen logging
-        txt  == plain text file logging
-        md   == markdown file logging
-        '''
-        pass
-
-    def GetPluginSettings(self):
+    def GetPluginSettings(self) -> Dict():
```
Comment (matthewfcarlson) [inline edk2toolext/invocables/edk2_ci_build.py:33]: Dict[str, Any]

File: edk2toolext/edk2_invocable.py (hunk 4 above)
Comment (spbrogan) [reply edk2toolext/edk2_invocable.py:94]: In Edk2Invocable we are doing tuple concatenation so i need a tuple. I tried to find out what would happen if I added a tuple to an iterable …but couldn't find anything good....so decided to just keep it as a tuple.

**Pull request merged**